bookwyrm/static/js/bookwyrm.js

```diff
@@ -0,0 +1,308 @@
+        }
+    }
+
+    /**
+     * Show or hide menus.
+     *
+     * @param  {Event} event
+     * @return {undefined}
+     */
+    toggleMenu(trigger, targetId) {
+        let expanded = trigger.getAttribute('aria-expanded') == 'false';
+
+        trigger.setAttribute('aria-expanded', expanded);
+
+        if (targetId) {
+            let target = document.getElementById(targetId);
+
+            this.addRemoveClass(target, 'is-active', expanded);
+        }
+    }
+
+    /**
+     * Show or hide generic containers.
+     *
+     * @param  {object}  container - DOM node
+     * @param  {boolean} pressed   - Is the trigger pressed?
+     * @return {undefined}
+     */
+    toggleContainer(container, pressed) {
+        this.addRemoveClass(container, 'is-hidden', pressed);
+    }
+
+    /**
+     * Check or uncheck a checbox.
+     *
+     * @param  {object}  checkbox - DOM node
+     * @param  {boolean} pressed  - Is the trigger pressed?
+     * @return {undefined}
+     */
+    toggleCheckbox(checkbox, pressed) {
+        document.getElementById(checkbox).checked = !!pressed;
+    }
+
+    /**
+     * Give the focus to an element.
+     * Only move the focus based on user interactions.
+     *
+     * @param  {string} nodeId - ID of the DOM node to focus (button, link…)
+     * @return {undefined}
+     */
+    toggleFocus(nodeId) {
+        let node = document.getElementById(nodeId);
+
+        node.focus();
+
+        setTimeout(function() {
+            node.selectionStart = node.selectionEnd = 10000;
+        }, 0);
+    }
+
+    /**
+     * Make a request and update the UI accordingly.
+     * This function is used for boosts and favourites.
+     *
+     * @param  {Event} event
+     * @return {undefined}
+     */
+    interact(event) {
+        event.preventDefault();
+
+        const bookwyrm = this;
+
+        let allTriggers = document.querySelectorAll(`.${event.target.dataset.id}`);
+
+        // Change icon to show ongoing activity on the current UI.
+        allTriggers.forEach(node => bookwyrm.addRemoveClass(
+            node,
+            'is-processing',
+            true
+        ));
+
+        this.ajaxPost(event.target)
+            .finally(() => {
+                // Change icon to remove ongoing activity on the current UI.
+                allTriggers.forEach(node => bookwyrm.addRemoveClass(
+                    node,
+                    'is-processing',
+                    false
+                ));
+            })
+            .then(function() {
+                allTriggers.forEach(node => bookwyrm.addRemoveClass(
+                    node,
+                    'is-hidden',
+                    node.className.indexOf('is-hidden') == -1
+                ));
+            })
+            .catch(error => {
+                // @todo Display a notification in the UI instead.
+                //       For now, the absence of change will be enough.
+                console.warn('Request failed:', error);
```

Comment:
unfortunately I think this will be an html response, so displaying it in the console won't be as helpful as it ought to be.

Reply:
That is the purpose of adding the [@todo](https://github.com/mouse-reeve/bookwyrm/search?q=todo), keeping this in consideration when implementing a global notification system, if ever. :)